### PR TITLE
refactor: use `//ui/shell_dialogs` on Linux

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/32857
Closes https://github.com/electron/electron/issues/41903
Refs https://github.com/electron/electron/pull/19159

This PR replaces our dialog implementation on Linux with upstream `//shell_dialogs`. This both streamlines and simplifies our dialog logic, and is ideally a first step towards using `//shell_dialogs` cross-platform. Per https://github.com/electron/electron/issues/32857#issuecomment-1683371233, this is the easiest to port over as it overlaps most completely with what's available upstream. By using Chromium's implementation, we also gain access to kde and portal-specific implementations of dialogs on Linux.

Notes:
- We should deprecate `showOverwriteConfirmation` as @aiddya noted in the chart comment, but that should be done in a discrete PR (it's the default on GTK 4+)

cc @bpasero

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

**Comment by ckerr** [comment]
@codebytere also noticed https://issues.chromium.org/issues/41469294#comment10, which says that upstream likely wants to support GtkFileChooserNative on GTK4

File: patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+index 53ae15f14c45ee72abdae172fc4555c9e4b3ff9a..9474e52c736b60c2060a359249c31b838e70a5c2 100644
+--- a/ui/gtk/select_file_dialog_linux_gtk.h
++++ b/ui/gtk/select_file_dialog_linux_gtk.h
+@@ -15,6 +15,9 @@
+ 
+ namespace gtk {
+ 
++// [label, show_overwrite_confirmation, show_hidden]
++typedef std::tuple<std::string, bool, bool> ExtraSettings;
++
```
Comment (ckerr) [inline patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch:195]: I would replace this with a helper struct:

```c++
struct ExtraSettings {
  std::string button_label;
  bool show_overwrite_confirmation = true;
  bool show_hidden = false;
};
```

- this will make it easier to add/remove/change parameters in future
- it makes the code easier to read
- the default values are built-in here instead of in a `make_tuple()` call

File: shell/browser/ui/file_dialog_linux.cc [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+// [save_label, show_overwrite_confirmation, show_hidden, ]
+std::tuple<std::string, bool, bool> GetSettingsTuple(
+    const DialogSettings& settings) {
+  bool show_overwrite_confirmation =
+      settings.properties & SAVE_DIALOG_SHOW_OVERWRITE_CONFIRMATION;
+  bool show_hidden = settings.properties & SAVE_DIALOG_SHOW_HIDDEN_FILES;
+
+  return std::make_tuple(settings.button_label, show_overwrite_confirmation,
+                         show_hidden);
+}
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:54]: Better to use `gtk::ExtraSettings` here than a tuple, since that way if `ExtraSettings` ever changes in `gtk/` then this code won't break.

For the same reason, I'd name this `GetExtraSettings()` to remove the reference to "tuple" :smile_cat: 

```c++
gtk::ExtraSettings GetExtraSettings(const DialogSettings& settings) {
  gtk::ExtraSettings extra;
  extra.button_label = settings.button_label;
  extra.show_overwrite_confirmation = settings.properties & SAVE_DIALOG_SHOW_OVERWRITE_CONFIRMATION;
  extra.show_hidden = settings.properties & SAVE_DIALOG_SHOW_HIDDEN_FILES;
  return extra;
}
```

^ this snippet assumes the other suggestion of making `ExtraSettings` a struct, but IMO the function name and return type change is worthwhile whether `ExtraSettings` is a struct or tuple

File: shell/browser/ui/file_dialog_linux.cc [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    callback_ = std::move(callback);
+    RunSaveDialogImpl(settings);
+  }
+
+  void RunOpenDialogImpl(const DialogSettings& settings) {
+    type_ = DialogType::OPEN;
+    ui::SelectFileDialog::FileTypeInfo file_info =
+        GetFilterInfo(settings.filters);
+    std::tuple<std::string, bool, bool> extra_settings =
+        GetSettingsTuple(settings);
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:124]:
```suggestion
    auto extra_settings = GetExtraSettings(settings);
```

File: shell/browser/ui/file_dialog_linux.cc [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  ~FileChooserDialog() override {
+    // TODO(codebytere): anything we need to do here?
+  }
+
+  void RunSaveDialogImpl(const DialogSettings& settings) {
+    type_ = DialogType::SAVE;
+    ui::SelectFileDialog::FileTypeInfo file_info =
+        GetFilterInfo(settings.filters);
+    std::tuple<std::string, bool, bool> extra_settings =
+        GetSettingsTuple(settings);
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:95]:
```suggestion
    auto extra_settings = GetExtraSettings(settings);
```

_(opinion)_  the function being called is visible only in this file, so I'd use `auto` here to futureproof this type.

File: shell/browser/ui/file_dialog_linux.cc [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+  return std::make_tuple(settings.button_label, show_overwrite_confirmation,
+                         show_hidden);
+}
+
+ui::SelectFileDialog::FileTypeInfo GetFilterInfo(const Filters& filters) {
+  ui::SelectFileDialog::FileTypeInfo file_type_info;
+
+  for (const auto& filter : filters) {
+    auto [name, extension_group] = filter;
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:60]:
```suggestion
    const auto& [name, extension_group] = filter;
```

File: shell/browser/ui/file_dialog_linux.cc [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    file_type_info.extension_description_overrides.push_back(
+        base::UTF8ToUTF16(name));
+
+    bool has_all_files_wildcard =
+        std::any_of(extension_group.begin(), extension_group.end(),
+                    [](std::string ext) { return ext == "*"; });
+    if (has_all_files_wildcard) {
+      file_type_info.include_all_files = true;
+    } else {
+      file_type_info.extensions.push_back(extension_group);
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:70]:
```suggestion
      file_type_info.extensions.emplace_back(extension_group);
```

File: shell/browser/ui/file_dialog_linux.cc [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        result.Get("filePaths", file_paths);
+        std::move(cb).Run();
+      },
+      run_loop.QuitClosure(), paths);
+
+  FileChooserDialog* dialog = new FileChooserDialog();
+  dialog->RunOpenDialog(std::move(cb), settings);
+
+  run_loop.Run();
+  return paths->size() > 0;
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:227]:
```suggestion
  return !paths->empty();
```

File: patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+index 698f97b23f851c02af037880585c82d4494da63f..2455d19fc4c470b640825922c04bfa71adf9bea7 100644
+--- a/ui/gtk/select_file_dialog_linux_gtk.cc
++++ b/ui/gtk/select_file_dialog_linux_gtk.cc
+@@ -243,6 +243,10 @@ void SelectFileDialogLinuxGtk::SelectFileImpl(
+ 
+   std::string title_string = base::UTF16ToUTF8(title);
+ 
++  ExtraSettings extra_settings = std::make_tuple("", true, false);
++  if (params)
++    extra_settings = *(reinterpret_cast<ExtraSettings*>(params));
```
Comment (ckerr) [inline patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch:23]:
```suggestion
+    extra_settings = *static_cast<ExtraSettings*>(params);
```

`void*` is a special case, use `static_cast` there instead of `reinterpret_cast`

File: patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+ // Implementation of SelectFileDialog that shows a Gtk common dialog for
+ // choosing a file or folder. This acts as a modal dialog.
+ class SelectFileDialogLinuxGtk : public ui::SelectFileDialogLinux,
+@@ -90,19 +93,23 @@ class SelectFileDialogLinuxGtk : public ui::SelectFileDialogLinux,
+   GtkWidget* CreateSelectFolderDialog(Type type,
+                                       const std::string& title,
+                                       const base::FilePath& default_path,
+-                                      gfx::NativeWindow parent);
++                                      gfx::NativeWindow parent,
++                                      ExtraSettings settings);
```
Comment (ckerr) [inline patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch:205]: Here and below in the other three methods:

```suggestion
+                                      const ExtraSettings& settings);
```

File: shell/browser/ui/file_dialog_linux.cc [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+ui::SelectFileDialog::Type GetDialogType(int properties) {
+  if (properties & OPEN_DIALOG_OPEN_FILE) {
+    return ui::SelectFileDialog::SELECT_OPEN_FILE;
+  } else if (properties & OPEN_DIALOG_OPEN_DIRECTORY) {
+    return ui::SelectFileDialog::SELECT_FOLDER;
+  } else if (properties & OPEN_DIALOG_MULTI_SELECTIONS) {
+    return ui::SelectFileDialog::SELECT_OPEN_MULTI_FILE;
+  }
+
+  return ui::SelectFileDialog::SELECT_OPEN_FILE;
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:42]: _(opinion)_ `else` after a return is a [little weird](https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html).

```suggestion
  if (properties & OPEN_DIALOG_OPEN_FILE)
    return ui::SelectFileDialog::SELECT_OPEN_FILE;
    
  if (properties & OPEN_DIALOG_OPEN_DIRECTORY)
    return ui::SelectFileDialog::SELECT_FOLDER;
    
  if (properties & OPEN_DIALOG_MULTI_SELECTIONS)
    return ui::SelectFileDialog::SELECT_OPEN_MULTI_FILE;
    
  return ui::SelectFileDialog::SELECT_OPEN_FILE;
```

File: shell/browser/ui/file_dialog_linux.cc [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+  ui::SelectFileDialog::FileTypeInfo file_type_info;
+
+  for (const auto& filter : filters) {
+    auto [name, extension_group] = filter;
+    file_type_info.extension_description_overrides.push_back(
+        base::UTF8ToUTF16(name));
+
+    bool has_all_files_wildcard =
+        std::any_of(extension_group.begin(), extension_group.end(),
+                    [](std::string ext) { return ext == "*"; });
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:66]:
```suggestion
    const bool has_all_files_wildcard = base::ranges::any_of(
        extension_group, [](const auto& ext) { return ext == "*"; });
```

**Review by ckerr — COMMENTED** [review]
Great idea, I love it. Especially with upstream indicating they will support the native dialogs in GTK4.

I have some implementation suggestions here

File: shell/browser/ui/file_dialog_linux.cc (hunk 5 above)
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:60]:
```suggestion
  for (const auto& [name, extension_group] : filters) {
```

File: shell/browser/ui/file_dialog_linux.cc [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+    auto dict = gin_helper::Dictionary::CreateEmpty(isolate);
+    dict.Set("canceled", false);
+    if (type_ == DialogType::SAVE) {
+      dict.Set("filePath", file.file_path);
+    } else {
+      dict.Set("filePaths", std::vector<base::FilePath>{file.file_path});
+    }
+
+    if (callback_) {
+      std::move(callback_).Run(dict);
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:161]:
```suggestion
      std::move(callback_).Run(std::move(dict));
```

File: shell/browser/ui/file_dialog_linux.cc [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+  void MultiFilesSelected(const std::vector<ui::SelectedFileInfo>& files,
+                          void* params) override {
+    v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+    v8::HandleScope scope(isolate);
+    auto dict = gin_helper::Dictionary::CreateEmpty(isolate);
+    dict.Set("canceled", false);
+    dict.Set("filePaths", ui::SelectedFileInfoListToFilePathList(files));
+
+    if (callback_) {
+      std::move(callback_).Run(dict);
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:176]:
```suggestion
      std::move(callback_).Run(std::move(dict));
```

Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc:176]: @ckerr this is true for `base::Value::Dict`s, but not `gin_helper::Dictionary`s - we don't need to move this one!

File: shell/browser/ui/file_dialog_linux.cc (hunk 12 above)
Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc:161]: @ckerr this is true for `base::Value::Dict`s, but not `gin_helper::Dictionary`s - we don't need to move this one!

File: shell/browser/ui/file_dialog_linux.cc (hunk 13 above)
Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc:176]: see above

File: shell/browser/ui/file_dialog_linux.cc (hunk 12 above)
Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc:157]: `base::Value::Dict` is move-only but `gin::Dictionary` isn't so i don't think we need this here 🤔

File: shell/browser/ui/file_dialog_linux.cc (hunk 13 above)
Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc:172]: same as above

File: shell/browser/ui/file_dialog_linux.cc (hunk 12 above)
Comment (ckerr) [reply shell/browser/ui/file_dialog_linux.cc:162]: You've got a point.

Usually move semantics are good when we build some nontrivial object and our final act is to ship it off to someone else, since it avoids a complex copy. But we don't have `Dictionary(Dictionary&&)` or `Dictionary(Dictionaryh)` -- only `Dictionary(const Dictionary& that)` -- so the move is moot here :shrug:

**Review by ckerr — APPROVED** [review]
LGTM.

File: shell/browser/ui/file_dialog_linux.cc [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+    } else {
+      promise_->Resolve(dict);
+    }
+  }
+
+ private:
+  DialogType type_;
+  scoped_refptr<ui::SelectFileDialog> dialog_;
+  base::OnceCallback<void(gin_helper::Dictionary)> callback_;
+  std::unique_ptr<gin_helper::Promise<gin_helper::Dictionary>> promise_;
```
Comment (deepak1556) [inline shell/browser/ui/file_dialog_linux.cc:205]: Any reason to hold the promise with a smart pointer ? Can this be just `gin_helper::Promise<gin_helper::Dictionary> promise_`

File: shell/browser/ui/file_dialog_linux.cc [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+    } else {
+      dict.Set("filePaths", std::vector<base::FilePath>{file.file_path});
+    }
+
+    if (callback_) {
+      std::move(callback_).Run(dict);
+    } else {
+      promise_->Resolve(dict);
+    }
+  }
```
Comment (deepak1556) [inline shell/browser/ui/file_dialog_linux.cc:166]: Missing `delete this` ? Similarly in other callbacks `MultiFilesSelected`, `FileSelectionCanceled` otherwise we are leaking the class on every dialog call.

File: shell/browser/ui/select_file_policy.cc [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    : web_contents_(web_contents) {}
+
+ElectronSelectFilePolicy::~ElectronSelectFilePolicy() = default;
+
+bool ElectronSelectFilePolicy::CanOpenSelectFileDialog() {
+  auto* web_preferences = electron::WebContentsPreferences::From(web_contents_);
+  if (!web_preferences)
+    return true;
+
+  return !web_preferences->ShouldDisableDialogs();
```
Comment (deepak1556) [inline shell/browser/ui/select_file_policy.cc:21]: `disableDialogs` options is not meant for file dialogs so relying on that value here seems incorrect. Also, `SelectFilePolicy` is optional in the upstream code, can we just set a `nullptr` during dialog creation if we are not planning to add a new file dialog disabling web preference ?

File: shell/browser/ui/file_dialog_gtk.cc [hunk 17]
```diff
@@ -1,395 +0,0 @@
...
-#include <memory>
-#include <string>
-
-#include "base/files/file_util.h"
-#include "base/functional/bind.h"
-#include "base/functional/callback.h"
-#include "base/memory/raw_ptr.h"
-#include "base/memory/raw_ptr_exclusion.h"
-#include "base/strings/string_util.h"
-#include "electron/electron_gtk_stubs.h"
```
Comment (deepak1556) [inline shell/browser/ui/file_dialog_gtk.cc:14]: Can also remove `https://github.com/electron/electron/blob/main/shell/browser/ui/electron_gtk.sigs` as part of this change.

**Review by deepak1556 — COMMENTED** [review]
This is great, left some minor feedback

File: shell/browser/ui/file_dialog_linux.cc [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+namespace file_dialog {
+
+DialogSettings::DialogSettings() = default;
+DialogSettings::DialogSettings(const DialogSettings&) = default;
+DialogSettings::~DialogSettings() = default;
+
+namespace {
+
+ui::SelectFileDialog::Type GetDialogType(int properties) {
+  if (properties & OPEN_DIALOG_OPEN_FILE)
```
Comment (aiddya) [inline shell/browser/ui/file_dialog_linux.cc:35]: The order appears to be incorrect. If both `OPEN_DIALOG_OPEN_FILE` and `OPEN_DIALOG_MULTI_SELECTIONS` are set, `SELECT_OPEN_MULTI_FILE` should be returned. This would return `SELECT_OPEN_FILE`.

File: shell/browser/ui/file_dialog_linux.cc [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+    return ui::SelectFileDialog::SELECT_OPEN_FILE;
+
+  if (properties & OPEN_DIALOG_OPEN_DIRECTORY)
+    return ui::SelectFileDialog::SELECT_FOLDER;
+
+  if (properties & OPEN_DIALOG_MULTI_SELECTIONS)
```
Comment (aiddya) [inline shell/browser/ui/file_dialog_linux.cc:41]: The earlier behavior was to use `GTK_FILE_CHOOSER_ACTION_SELECT_FOLDER` for directories and `GTK_FILE_CHOOSER_ACTION_OPEN` for files and setting `gtk_file_chooser_set_select_multiple` if multi-selections bit was set.

`SELECT_OPEN_MULTI_FILE` would result in a call to `CreateMultiFileOpenDialog()` which always uses `GTK_FILE_CHOOSER_ACTION_OPEN`. Can multiple directories be selected with this?

File: shell/browser/ui/file_dialog_linux.cc [hunk 20]
```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2024 Microsoft, Gmbh.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include <memory>
```
Comment (aiddya) [inline shell/browser/ui/file_dialog_linux.cc:5]: Can you check if `shell_dialog_linux::Initialize()` is called during or soon after Electron initialization? If not, it should be called sometime early so that portal availability can be verified. This would avoid delays while opening the first dialog on systems that don't support portals.

**Review by aiddya — COMMENTED** [review]
Thanks for working on this!

File: shell/browser/ui/file_dialog_linux.cc (hunk 20 above)
Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc:5]: @aiddya yes! `shell_dialog_linux::Initialize()` is called by https://github.com/electron/electron/blob/b428315c6dd9aea97611ab3ef98145b870ef33ae/shell/browser/electron_browser_main_parts.cc#L397

File: shell/browser/ui/select_file_policy.cc (hunk 16 above)
Comment (codebytere) [reply shell/browser/ui/select_file_policy.cc]: yes, removed :)

File: shell/browser/ui/file_dialog_linux.cc (hunk 19 above)
Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc:40]: ah - i see the issue. I'll push a fix momentarily!

Comment (codebytere) [reply shell/browser/ui/file_dialog_linux.cc]: done in 255f3bf0e0

**Review by aiddya — COMMENTED** [review]
LGTM!

File: shell/browser/ui/file_dialog_linux.cc [hunk 21]
```diff
@@ -31,15 +31,15 @@ DialogSettings::~DialogSettings() = default;
...
     return ui::SelectFileDialog::SELECT_OPEN_FILE;
+  }
 
   if (properties & OPEN_DIALOG_OPEN_DIRECTORY)
     return ui::SelectFileDialog::SELECT_FOLDER;
 
-  if (properties & OPEN_DIALOG_MULTI_SELECTIONS)
-    return ui::SelectFileDialog::SELECT_OPEN_MULTI_FILE;
-
   return ui::SelectFileDialog::SELECT_OPEN_FILE;
```
Comment (ckerr) [inline shell/browser/ui/file_dialog_linux.cc:43]: These conditionals + `SELECT_OPEN_FILE` being returned from two different logic branches is really hard to follow.

Would this be equivalent?

```c++
if (properties & OPEN_DIALOG_OPEN_DIRECTORY)
  return ui::SelectFileDialog::SELECT_FOLDER;

if (properties & OPEN_DIALOG_MULTI_SELECTIONS)
  return ui::SelectFileDialog::SELECT_OPEN_MULTI_FILE;

return ui::SelectFileDialog::SELECT_OPEN_FILE;
```

**Review by deepak1556 — APPROVED** [review]
👏

**Comment by drankinatty** [comment]
If this is the change where you search for command line utilities to use as your file-chooser, like `kdialog`, etc.., this breaks fileChoose display on Trinity Desktop Environment TDE/KDE3. It further breaks any chooser where multiple versions of a chooser like `kdialog` are installed and the `$PATH` ordering places the older first.

Any fileChooser scheme should not depend on how the `$PATH` variable may or may not be set and should not utilize command-line utilities to generate fileChoosers without validating the version of the utility and whether or not it supports the options it is sent. Ultimately, this breakage is one of basic validation and providing a fallback that was not done in this change. If the fileChooser selected does not display and instead throws an error, then that needs to be handled and the default Gtk choosers need to be shown as they were up until vscode ver: 1.89. 

See the bug linked by @daiyam above [Linux - vscodium no longer displayes the file-dialog after update to Ver. 1.9.0](https://github.com/VSCodium/vscodium/issues/1927). This affects more than just TDE/KDE3 as noted by the comments in that bug report. It also effects vscode/vscodium identically. Several work-arounds (ugly hacks) are included in the linked report that either involve intercepting the call to the command-line utility and rewriting the command or to alter the `$PATH` to remove the search-path to the utility forcing the use of the normal Gtk dialogs. Neither are realistic.

It may be convenient to simply search the Linux `PATH` for a set of executable names to use as fileChoosers, but there is no telling what you actually find. So there needs to be additional validation that the fileChoose found is actually what you think it is, and if not, provide the Gtk fallback.

Lastly, there needs to be a clear way to disable this new behavior and restore the default Gtk dialogs on desktop that this breaks fileChooser display on. Mozilla provides `widget.use-xdg-desktop-portal.file-picker` which when set to `0` disables searching for native chooser utilities and restores default Gtk dialog use. Something like that in the vscode preferences would work, or a command line option. Because, above all, a *New Feature* to one, is a *BUG* to another -- if it cannot be turned off.

Hopefully the behavior of this commit can be made more robust to prevent the dialog failures it causes. Absent that a clear option to disable it and restore default dialogs would be greatly appreciated.
